[aotinductor] add versions for the sdpa shim api

chenyang78 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113487

In our first implemenation of the sdpa shim api, we didn't consider
the case where the optional scale argument could be None. It was
unnoticed because we always got a default argument for the cuda backend.
The issue was detected with the cpu backend.

This PR implements versioning for shim kernels. Currently, we only
have different versions for the sdpa api. We expect we would only
maintain a very small number of abi-compatible shim APIs that
had different versions.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @kadeng @muchulee8 @aakhundov @ColinPeppler

Commit:
[aotinductor] add versions for the sdpa shim api

In our first implemenation of the sdpa shim api, we didn't consider
the case where the optional scale argument could be None. It was
unnoticed because we always got a default argument for the cuda backend.
The issue was detected with the cpu backend.

This PR implements versioning for shim kernels. Currently, we only
have different versions for the sdpa api. We expect we would only
maintain a very small number of abi-compatible shim APIs that
had different versions.

[ghstack-poisoned]

int3 commented on torch/_inductor/ir.py:
```diff
@@ -3777,9 +3775,35 @@ def __init__(self, count: int, device: torch.device):
+            return self.kernel
+
+        def sdpa_ver_fn():
+            # For sdpa, we need the v2 version only if any optional
+            # kwarg is missing.
```
I thought the idea was to make all new models call the v2 version, so that we could delete the old one entirely? Doesn't using the v2 version only if an optional arg is missing make migration harder?

chenyang78 replied:
> I thought the idea was to make all new models call the v2 version, so that we could delete the old one entirely? Doesn't using the v2 version only if an optional arg is missing make migration harder?

It's not about new models v.s. old models. Given the same model, the optional scale could be either None or have a default value, for different backends (e.g. cpu and cuda in our case). Regarding migrating to the V2 version, my understanding is that once the new V2 version becomes available on the serving side, we could always generate the V2 version. At that point, we may safely remove the old one.

int3 replied:
Ohh right. We want to keep generating calls to the old (v1) version where possible for now so the CUDA models keep working with old libtorch deployments. Got it

int3 commented on torch/_inductor/ir.py:
```diff
@@ -4278,6 +4302,7 @@ def __init__(
                 self.kernel = (
                     f"{kernel.__module__.replace('._ops.', '.ops.')}.{kernel.__name__}"
                 )
+        self.abi_compatible_kernel = None
```
not sure why the typechecker didn't complain, but shouldn't this line be in the constructor of `ExternKernelAlloc`, since `ExternKernelAlloc.codegen` references it?

chenyang78 replied:
Good catch! Fixed. Thanks.

Commit:
Update on "[aotinductor] add versions for the sdpa shim api"


In our first implemenation of the sdpa shim api, we didn't consider
the case where the optional scale argument could be None. It was
unnoticed because we always got a default argument for the cuda backend.
The issue was detected with the cpu backend.

This PR implements versioning for shim kernels. Currently, we only
have different versions for the sdpa api. We expect we would only
maintain a very small number of abi-compatible shim APIs that
had different versions.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -196,6 +197,26 @@ AOTI_TORCH_EXPORT AOTITorchError aoti_torch__scaled_dot_product_flash_attention(
+    AtenTensorHandle key,
+    AtenTensorHandle value,
+    double dropout_p,
+    bool is_causal,
```
I haven't paid close enough attention to the v1 function signature, but let's use int instead of bool in the v2 function signature to avoid any potential problems from mixing c++ compilers.

chenyang78 replied:
Good point. Fixed. Thanks.

desertfire reviewed (changes requested):
Left a comment about the function signature. The rest LGTM.

Commit:
Update on "[aotinductor] add versions for the sdpa shim api"


In our first implemenation of the sdpa shim api, we didn't consider
the case where the optional scale argument could be None. It was
unnoticed because we always got a default argument for the cuda backend.
The issue was detected with the cpu backend.

This PR implements versioning for shim kernels. Currently, we only
have different versions for the sdpa api. We expect we would only
maintain a very small number of abi-compatible shim APIs that
had different versions.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

chenyang78 commented:
@pytorchbot merge
